packages/core/src/web/bridges/mcp-app-adapter.ts

```diff
@@ -5,9 +5,49 @@ import type {
   McpUiRequestDisplayModeRequest,
   McpUiRequestDisplayModeResult,
 } from "@modelcontextprotocol/ext-apps";
+import type {
+  CallToolRequest,
+  CallToolResult,
+} from "@modelcontextprotocol/sdk/types.js";
+
 import type { BridgeExternalStore } from "./hooks/types.js";
 import { McpAppBridge } from "./mcp-app-bridge.js";
-import type { BridgeInterface, Methods } from "./types.js";
+import type { BridgeInterface, CallToolResponse, Methods } from "./types.js";
+
+export const callTool = async <
+  ToolArgs extends Record<string, unknown> | null = null,
+  ToolResponse extends CallToolResponse = CallToolResponse,
+>(
+  name: string,
+  args: ToolArgs,
+): Promise<ToolResponse> => {
+  const bridge = McpAppBridge.getInstance();
+  const response = await bridge.request<CallToolRequest, CallToolResult>({
+    method: "tools/call",
+    params: {
+      name,
+      arguments: args ?? undefined,
+    },
+  });
+
+  const text: string[] = [];
+
+  response.content.forEach((content) => {
+    if (content.type === "text") {
+      text.push(content.text);
+    }
+  });
+
+  const result = text.join("\n");
+
+  return {
+    content: response.content,
+    structuredContent: response.structuredContent ?? {},
+    isError: response.isError ?? false,
+    result,
+    meta: response._meta ?? {},
+  } as ToolResponse;
```

Comment:
Why do you need typecasting here?

Reply:
Well it's something we can see in the implementation - calling a tool is like does a fetch call (external) so I cannot ensure programmatically that the response will be of this type.

I could add a type hint directly on the callTool Request. But for now it don't think it's worth it.

WDYT ?